modules/images/image-loading-optimization/storage/lock.php

```diff
@@ -0,0 +1,69 @@
+<?php
+/**
+ * Metrics storage lock.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Gets the TTL (in seconds) for the page metric storage lock.
+ *
+ * @return int TTL in seconds.
+ */
+function ilo_get_page_metric_storage_lock_ttl(): int {
+
+	/**
+	 * Filters how long a given IP is locked from submitting another metric-storage REST API request.
+	 *
+	 * Filtering the TTL to zero will disable any metric storage locking. This is useful during development.
+	 *
+	 * @param int $ttl TTL.
+	 */
+	return (int) apply_filters( 'ilo_metrics_storage_lock_ttl', MINUTE_IN_SECONDS );
```

Comment:
Maybe we should document that this can also be `0` and that, in that case, transients shouldn't be used at all?

Two related thoughts:
* What is the rationale for allowing `0`?
* If we keep it, we need to triple-check that this value is never simply passed through to `set_transient()`. As far as I can tell, that's currently (rightfully) not the case, but it's a bit error-prone, and in case of an error it would lead to transients without expiration, which ties into my other scalability concern with lots of transients. With expiring transients, a large number can be fine, but not if they remain forever.

I think to harden against the potential problems from my second point, it would be good to make this a private function, so that nobody can call it externally (potentially without knowing these implications).

I think the 4 functions in this file could be easily refactored into e.g. a `ILO_Page_Metric_Storage_Lock` class, where the methods to get the TTL and transient key could be private. The public interface could then simply be the `lock()` and `is_locked()` methods.

Alternatively, we disallow returning `0` via the filter, avoiding the risk in a simple way. We could also reconsider later, combining with potential refactoring to keep use of this function "contained".

Reply:
Yes, it is useful to allow zero, beyond just for development purposes. For example, consider this example:

```php
add_filter( 'ilo_metrics_storage_lock_ttl', static function ( $ttl ) {
    return is_user_logged_in() ? 0 : $ttl;
} );
```

This allows authenticated users in WP to store page metrics without any locking, which could even make sense to do by default. A user who is already authenticated in WP is not going to be the same level of risk to necessitate locking. I've added this as an example in the filter doc in da0b4208b73694d1e1800cb04178140b9837b094.

Also, I'm now ensuring that the TTL is never less than zero.  

As for when the TTL is zero exactly, code is in place to make sure that no transient is set:

https://github.com/WordPress/performance/blob/da0b4208b73694d1e1800cb04178140b9837b094/modules/images/image-loading-optimization/storage/lock.php#L53-L61

So there is no possible way for a transient to be set with a zero expiration. Not really a need to make it private, though I intend to mark all of the functions as private in any case.

Reply:
Well, making the function private ensures nobody uses it incorrectly. Obviously we can't truly make it private as long as it's a function, but if someone still uses it, we have no obligation to keep it working for them.

The rationale for this specific function to be private is that it's error-prone to _have_ to remember that you have to always add an extra condition for whether TTL is `0` since otherwise you'd create transients without expiration. That's why I think this function specifically really should be private, or extremely well documented on what should happen in the case it returns `0`. See https://github.com/WordPress/performance/pull/878#discussion_r1394643420.